Stdlib.flush hangs when used within signal handler

I want to flush output channels periodically.

If I use `flush stdout` in *both* the signal handler *and* the regular code, the application hangs.

Here a minimal reproducible example (which I tried on Linux and MacOS using OCaml 4.08 and 4.10):
<pre>
let n_flush = ref 0;;

let main () =
  Sys.set_signal Sys.sigalrm (Signal_handle (fun _ -> incr n_flush; flush stdout));
  ignore(Unix.setitimer Unix.ITIMER_REAL { it_value = 0.001; it_interval = 0.001 });

  for i = 1 to 100 do
    let s = string_of_int i in
    print_string s;
    flush stdout;
    Unix.sleepf 0.02;
    for j = 1 to String.length s do print_char '\b' done
  done;
  print_string ("n_flush = " ^ string_of_int !n_flush);
  print_endline " ... DONE!";;

main ();;
</pre>

Note that above time intervals are a lot higher in practise (~1/10s), I pumped them up so that the issue shows more easily.

<!--
Welcome to OCaml's Issue tracker!

OCaml's developers use this tracker for bugs and feature requests only, rather
than user support.

If you have questions about *using* OCaml, please ask at
https://discuss.ocaml.org (more people read Discuss than this tracker, and
you'll get confirmation of whether you've really found a bug or need a new
feature).

If your error came from the OCaml package manager, opam, (messages beginning
`[ERROR] The compilation of ...`), please start at
https://github.com/ocaml/opam-repository/issues/new.

Some libraries and tools which used to be part of OCaml are now maintained
separately. Please post questions about Graphics, Num, camlp4, LablTk, CamlDBM
or OCamlTopWin on Discuss or on their respective issue trackers (see [README.adoc](https://github.com/ocaml/ocaml/blob/trunk/README.adoc#separately-maintained-components)
for a full list).
-->

--- a/Changes
+++ b/Changes
@@ -94,6 +94,16 @@ Working version
    when using mutually recursive functions.
    (Jacques-Henri Jourdan, review François Bobot)
 
+* #1128, #7503, #9036, #9722: EINTR-based signal handling.
+  When a signal arrives, avoid running its OCaml handler in the middle
+  of a blocking section. Instead, allow control to return quickly to
+  a polling point where the signal handler can safely run, ensuring that
+  I/O locks are not held while it runs. A polling point was removed from
+  caml_leave_blocking_section, and one added to caml_raise.
+  (Stephen Dolan, review by Goswin von Brederlow, Xavier Leroy, Damien
+   Doligez, Anil Madhavapeddy, Guillaume Munch-Maccagnoni and Jacques-
+   Henri Jourdan)
+
 ### Code generation and optimizations:
 
 - #9551: ocamlc no longer loads DLLs at link time to check that

--- a/ocamltest/run_stubs.c
+++ b/ocamltest/run_stubs.c
@@ -71,8 +71,10 @@ static void logToChannel(void *voidchannel, const char *fmt, va_list ap)
     if (text == NULL) return;
     if (vsnprintf(text, length, fmt, ap) != length) goto end;
   }
+  Lock(channel);
   caml_putblock(channel, text, length);
   caml_flush(channel);
+  Unlock(channel);
 end:
   free(text);
 }

--- a/otherlibs/systhreads/st_posix.h
+++ b/otherlibs/systhreads/st_posix.h
@@ -437,6 +437,8 @@ value caml_thread_sigmask(value cmd, value sigs) /* ML */
   retcode = pthread_sigmask(how, &set, &oldset);
   caml_leave_blocking_section();
   st_check_error(retcode, "Thread.sigmask");
+  /* Run any handlers for just-unmasked pending signals */
+  caml_process_pending_actions();
   return st_encode_sigset(&oldset);
 }
 

--- a/otherlibs/systhreads/st_stubs.c
+++ b/otherlibs/systhreads/st_stubs.c
@@ -253,15 +253,6 @@ static void caml_thread_leave_blocking_section(void)
   caml_thread_restore_runtime_state();
 }
 
-static int caml_thread_try_leave_blocking_section(void)
-{
-  /* Disable immediate processing of signals (PR#3659).
-     try_leave_blocking_section always fails, forcing the signal to be
-     recorded and processed at the next leave_blocking_section or
-     polling. */
-  return 0;
-}
-
 /* Hooks for I/O locking */
 
 static void caml_io_mutex_free(struct channel *chan)
@@ -496,7 +487,6 @@ CAMLprim value caml_thread_initialize(value unit)   /* ML */
   caml_scan_roots_hook = caml_thread_scan_roots;
   caml_enter_blocking_section_hook = caml_thread_enter_blocking_section;
   caml_leave_blocking_section_hook = caml_thread_leave_blocking_section;
-  caml_try_leave_blocking_section_hook = caml_thread_try_leave_blocking_section;
 #ifdef NATIVE_CODE
   caml_termination_hook = st_thread_exit;
 #endif

--- a/otherlibs/unix/signals.c
+++ b/otherlibs/unix/signals.c
@@ -71,6 +71,8 @@ CAMLprim value unix_sigprocmask(value vaction, value vset)
   caml_enter_blocking_section();
   retcode = caml_sigmask_hook(how, &set, &oldset);
   caml_leave_blocking_section();
+  /* Run any handlers for just-unmasked pending signals */
+  caml_process_pending_actions();
   if (retcode != 0) unix_error(retcode, "sigprocmask", Nothing);
   return encode_sigset(&oldset);
 }

--- a/runtime/backtrace_byt.c
+++ b/runtime/backtrace_byt.c
@@ -386,6 +386,7 @@ static void read_main_debug_info(struct debug_info *di)
   if (caml_seek_optional_section(fd, &trail, "DBUG") != -1) {
     chan = caml_open_descriptor_in(fd);
 
+    Lock(chan);
     num_events = caml_getword(chan);
     events = caml_alloc(num_events, 0);
 
@@ -401,6 +402,7 @@ static void read_main_debug_info(struct debug_info *di)
       /* Record event list */
       Store_field(events, i, evl);
     }
+    Unlock(chan);
 
     caml_close_channel(chan);
 

--- a/runtime/caml/compatibility.h
+++ b/runtime/caml/compatibility.h
@@ -264,7 +264,6 @@
 #define something_to_do caml_something_to_do
 #define enter_blocking_section_hook caml_enter_blocking_section_hook
 #define leave_blocking_section_hook caml_leave_blocking_section_hook
-#define try_leave_blocking_section_hook caml_try_leave_blocking_section_hook
 #define enter_blocking_section caml_enter_blocking_section
 #define leave_blocking_section caml_leave_blocking_section
 #define convert_signal_number caml_convert_signal_number

--- a/runtime/caml/io.h
+++ b/runtime/caml/io.h
@@ -64,8 +64,15 @@ enum {
      [offset] is the absolute position of the logical end of the buffer, [max].
 */
 
-/* Functions and macros that can be called from C.  Take arguments of
-   type struct channel *.  The channel must be locked before calling these. */
+/* Creating and closing channels from C */
+
+CAMLextern struct channel * caml_open_descriptor_in (int);
+CAMLextern struct channel * caml_open_descriptor_out (int);
+CAMLextern void caml_close_channel (struct channel *);
+
+
+/* I/O on channels from C. The channel must be locked (see below) before
+   calling any of the functions and macros below */
 
 #define caml_putch(channel, ch) do{                                       \
   if ((channel)->curr >= (channel)->end) caml_flush_partial(channel);     \
@@ -77,11 +84,8 @@ enum {
    ? caml_refill(channel)                                                   \
    : (unsigned char) *((channel)->curr)++)
 
-CAMLextern struct channel * caml_open_descriptor_in (int);
-CAMLextern struct channel * caml_open_descriptor_out (int);
-CAMLextern void caml_close_channel (struct channel *);
-CAMLextern int caml_channel_binary_mode (struct channel *);
 CAMLextern value caml_alloc_channel(struct channel *chan);
+CAMLextern int caml_channel_binary_mode (struct channel *);
 
 CAMLextern int caml_flush_partial (struct channel *);
 CAMLextern void caml_flush (struct channel *);

--- a/runtime/caml/io.h
+++ b/runtime/caml/io.h
@@ -65,7 +65,7 @@ enum {
 */
 
 /* Functions and macros that can be called from C.  Take arguments of
-   type struct channel *.  No locking is performed. */
+   type struct channel *.  The channel must be locked before calling these. */
 
 #define caml_putch(channel, ch) do{                                       \
   if ((channel)->curr >= (channel)->end) caml_flush_partial(channel);     \

--- a/runtime/caml/osdeps.h
+++ b/runtime/caml/osdeps.h
@@ -30,12 +30,16 @@ extern unsigned short caml_win32_revision;
 #include "misc.h"
 #include "memory.h"
 
+#define Io_interrupted (-1)
+
 /* Read at most [n] bytes from file descriptor [fd] into buffer [buf].
    [flags] indicates whether [fd] is a socket
    (bit [CHANNEL_FLAG_FROM_SOCKET] is set in this case, see [io.h]).
    (This distinction matters for Win32, but not for Unix.)
    Return number of bytes read.
-   In case of error, raises [Sys_error] or [Sys_blocked_io]. */
+   In case of error, raises [Sys_error] or [Sys_blocked_io].
+   If interrupted by a signal and no bytes where read, returns
+   Io_interrupted without raising. */
 extern int caml_read_fd(int fd, int flags, void * buf, int n);
 
 /* Write at most [n] bytes from buffer [buf] onto file descriptor [fd].
@@ -43,7 +47,9 @@ extern int caml_read_fd(int fd, int flags, void * buf, int n);
    (bit [CHANNEL_FLAG_FROM_SOCKET] is set in this case, see [io.h]).
    (This distinction matters for Win32, but not for Unix.)
    Return number of bytes written.
-   In case of error, raises [Sys_error] or [Sys_blocked_io]. */
+   In case of error, raises [Sys_error] or [Sys_blocked_io].
+   If interrupted by a signal and no bytes were written, returns
+   Io_interrupted without raising. */
 extern int caml_write_fd(int fd, int flags, void * buf, int n);
 
 /* Decompose the given path into a list of directories, and add them

--- a/runtime/caml/signals.h
+++ b/runtime/caml/signals.h
@@ -31,6 +31,7 @@ extern "C" {
 #endif
 
 CAMLextern void caml_enter_blocking_section (void);
+CAMLextern void caml_enter_blocking_section_no_pending (void);
 CAMLextern void caml_leave_blocking_section (void);
 
 CAMLextern void caml_process_pending_actions (void);
@@ -38,6 +39,9 @@ CAMLextern void caml_process_pending_actions (void);
    minor and major collections, signal handlers, finalisers, and
    Memprof callbacks. Assumes that the runtime lock is held. Can raise
    exceptions asynchronously into OCaml code. */
+
+CAMLextern int caml_check_pending_actions (void);
+/* Returns 1 if there are pending actions, 0 otherwise. */
 
 CAMLextern value caml_process_pending_actions_exn (void);
 /* Same as [caml_process_pending_actions], but returns the exception

--- a/runtime/caml/signals.h
+++ b/runtime/caml/signals.h
@@ -86,7 +86,6 @@ void caml_setup_stack_overflow_detection(void);
 
 CAMLextern void (*caml_enter_blocking_section_hook)(void);
 CAMLextern void (*caml_leave_blocking_section_hook)(void);
-CAMLextern int (*caml_try_leave_blocking_section_hook)(void);
 #ifdef POSIX_SIGNALS
 CAMLextern int (*caml_sigmask_hook)(int, const sigset_t *, sigset_t *);
 #endif

--- a/runtime/debugger.c
+++ b/runtime/debugger.c
@@ -141,6 +141,12 @@ static void open_connection(void)
 #endif
   dbg_in = caml_open_descriptor_in(dbg_socket);
   dbg_out = caml_open_descriptor_out(dbg_socket);
+  /* The code in this file does not bracket channel I/O operations with
+     Lock and Unlock, so fail if those are not no-ops. */
+  if (caml_channel_mutex_lock != NULL ||
+      caml_channel_mutex_unlock != NULL ||
+      caml_channel_mutex_unlock_exn != NULL)
+    caml_fatal_error("debugger does not support channel locks");
   if (!caml_debugger_in_use) caml_putword(dbg_out, -1); /* first connection */
 #ifdef _WIN32
   caml_putword(dbg_out, _getpid());

--- a/runtime/fail_byt.c
+++ b/runtime/fail_byt.c
@@ -34,6 +34,8 @@
 CAMLexport void caml_raise(value v)
 {
   Unlock_exn();
+  CAMLassert(!Is_exception_result(v));
+  v = caml_process_pending_actions_with_root(v);
   Caml_state->exn_bucket = v;
   if (Caml_state->external_raise == NULL) caml_fatal_uncaught_exception(v);
   siglongjmp(Caml_state->external_raise->buf, 1);

--- a/runtime/fail_nat.c
+++ b/runtime/fail_nat.c
@@ -62,6 +62,10 @@ CAMLno_asan
 void caml_raise(value v)
 {
   Unlock_exn();
+
+  CAMLassert(!Is_exception_result(v));
+  v = caml_process_pending_actions_with_root(v);
+
   if (Caml_state->exception_pointer == NULL) caml_fatal_uncaught_exception(v);
 
   while (Caml_state->local_roots != NULL &&

--- a/runtime/io.c
+++ b/runtime/io.c
@@ -69,13 +69,24 @@ CAMLexport struct channel * caml_all_opened_channels = NULL;
 
 /* Functions shared between input and output */
 
+static void check_pending(struct channel *channel)
+{
+  if (caml_check_pending_actions()) {
+    /* Temporarily unlock the channel, to ensure locks are not held
+       while any signal handlers (or finalisers, etc) are running */
+    Unlock(channel);
+    caml_process_pending_actions();
+    Lock(channel);
+  }
+}
+
 CAMLexport struct channel * caml_open_descriptor_in(int fd)
 {
   struct channel * channel;
 
   channel = (struct channel *) caml_stat_alloc(sizeof(struct channel));
   channel->fd = fd;
-  caml_enter_blocking_section();
+  caml_enter_blocking_section_no_pending();
   channel->offset = lseek(fd, 0, SEEK_CUR);
   caml_leave_blocking_section();
   channel->curr = channel->max = channel->buff;
@@ -131,12 +142,13 @@ CAMLexport file_offset caml_channel_size(struct channel *channel)
   file_offset offset;
   file_offset end;
   int fd;
+  check_pending(channel);
 
   /* We extract data from [channel] before dropping the OCaml lock, in case
      someone else touches the block. */
   fd = channel->fd;
   offset = channel->offset;
-  caml_enter_blocking_section();
+  caml_enter_blocking_section_no_pending();
   end = lseek(fd, 0, SEEK_END);
   if (end == -1 || lseek(fd, offset, SEEK_SET) != offset) {
     caml_leave_blocking_section();
@@ -167,12 +179,15 @@ CAMLexport int caml_channel_binary_mode(struct channel *channel)
 CAMLexport int caml_flush_partial(struct channel *channel)
 {
   int towrite, written;
+ again:
+  check_pending(channel);
 
   towrite = channel->curr - channel->buff;
   CAMLassert (towrite >= 0);
   if (towrite > 0) {
     written = caml_write_fd(channel->fd, channel->flags,
                             channel->buff, towrite);
+    if (written == Io_interrupted) goto again;
     channel->offset += written;
     if (written < towrite)
       memmove(channel->buff, channel->buff + written, towrite - written);
@@ -202,7 +217,7 @@ CAMLexport void caml_putword(struct channel *channel, uint32_t w)
 
 CAMLexport int caml_putblock(struct channel *channel, char *p, intnat len)
 {
-  int n, free, towrite, written;
+  int n, free;
 
   n = len >= INT_MAX ? INT_MAX : (int) len;
   free = channel->end - channel->curr;
@@ -215,13 +230,8 @@ CAMLexport int caml_putblock(struct channel *channel, char *p, intnat len)
     /* Write request overflows buffer (or just fills it up): transfer whatever
        fits to buffer and write the buffer */
     memmove(channel->curr, p, free);
-    towrite = channel->end - channel->buff;
-    written = caml_write_fd(channel->fd, channel->flags,
-                            channel->buff, towrite);
-    if (written < towrite)
-      memmove(channel->buff, channel->buff + written, towrite - written);
-    channel->offset += written;
-    channel->curr = channel->end - written;
+    channel->curr = channel->end;
+    caml_flush_partial(channel);
     return free;
   }
 }
@@ -240,7 +250,7 @@ CAMLexport void caml_really_putblock(struct channel *channel,
 CAMLexport void caml_seek_out(struct channel *channel, file_offset dest)
 {
   caml_flush(channel);
-  caml_enter_blocking_section();
+  caml_enter_blocking_section_no_pending();
   if (lseek(channel->fd, dest, SEEK_SET) != dest) {
     caml_leave_blocking_section();
     caml_sys_error(NO_ARG);
@@ -259,16 +269,22 @@ CAMLexport file_offset caml_pos_out(struct channel *channel)
 /* caml_do_read is exported for Cash */
 CAMLexport int caml_do_read(int fd, char *p, unsigned int n)
 {
-  return caml_read_fd(fd, 0, p, n);
+  int r;
+  do {
+    r = caml_read_fd(fd, 0, p, n);
+  } while (r == Io_interrupted);
+  return r;
 }
 
 CAMLexport unsigned char caml_refill(struct channel *channel)
 {
   int n;
-
+ again:
+  check_pending(channel);
   n = caml_read_fd(channel->fd, channel->flags,
                    channel->buff, channel->end - channel->buff);
-  if (n == 0) caml_raise_end_of_file();
+  if (n == Io_interrupted) goto again;
+  else if (n == 0) caml_raise_end_of_file();
   channel->offset += n;
   channel->max = channel->buff + n;
   channel->curr = channel->buff + 1;
@@ -292,7 +308,8 @@ CAMLexport uint32_t caml_getword(struct channel *channel)
 CAMLexport int caml_getblock(struct channel *channel, char *p, intnat len)
 {
   int n, avail, nread;
-
+ again:
+  check_pending(channel);
   n = len >= INT_MAX ? INT_MAX : (int) len;
   avail = channel->max - channel->curr;
   if (n <= avail) {
@@ -306,6 +323,7 @@ CAMLexport int caml_getblock(struct channel *channel, char *p, intnat len)
   } else {
     nread = caml_read_fd(channel->fd, channel->flags, channel->buff,
                          channel->end - channel->buff);
+    if (nread == Io_interrupted) goto again;
     channel->offset += nread;
     channel->max = channel->buff + nread;
     if (n > nread) n = nread;
@@ -335,7 +353,7 @@ CAMLexport void caml_seek_in(struct channel *channel, file_offset dest)
       dest <= channel->offset) {
     channel->curr = channel->max - (channel->offset - dest);
   } else {
-    caml_enter_blocking_section();
+    caml_enter_blocking_section_no_pending();
     if (lseek(channel->fd, dest, SEEK_SET) != dest) {
       caml_leave_blocking_section();
       caml_sys_error(NO_ARG);
@@ -355,7 +373,8 @@ CAMLexport intnat caml_input_scan_line(struct channel *channel)
 {
   char * p;
   int n;
-
+ again:
+  check_pending(channel);
   p = channel->curr;
   do {
     if (p >= channel->max) {
@@ -378,7 +397,8 @@ CAMLexport intnat caml_input_scan_line(struct channel *channel)
       /* Fill the buffer as much as possible */
       n = caml_read_fd(channel->fd, channel->flags,
                        channel->max, channel->end - channel->max);
-      if (n == 0) {
+      if (n == Io_interrupted) goto again;
+      else if (n == 0) {
         /* End-of-file encountered. Return the number of characters in the
            buffer, with negative sign since we haven't encountered
            a newline. */
@@ -545,7 +565,7 @@ CAMLprim value caml_ml_close_channel(value vchannel)
   channel->curr = channel->max = channel->end;
 
   if (do_syscall) {
-    caml_enter_blocking_section();
+    caml_enter_blocking_section_no_pending();
     result = close(fd);
     caml_leave_blocking_section();
   }
@@ -563,16 +583,27 @@ CAMLprim value caml_ml_close_channel(value vchannel)
 #define EOVERFLOW ERANGE
 #endif
 
+static file_offset ml_channel_size(value vchannel)
+{
+  CAMLparam1 (vchannel);
+  struct channel * channel = Channel(vchannel);
+  file_offset size;
+
+  Lock(channel);
+  size = caml_channel_size(Channel(vchannel));
+  Unlock(channel);
+  if (size > Max_long) { errno = EOVERFLOW; caml_sys_error(NO_ARG); }
+  CAMLreturnT(file_offset, size);
+}
+
 CAMLprim value caml_ml_channel_size(value vchannel)
 {
-  file_offset size = caml_channel_size(Channel(vchannel));
-  if (size > Max_long) { errno = EOVERFLOW; caml_sys_error(NO_ARG); }
-  return Val_long(size);
+  return Val_long(ml_channel_size(vchannel));
 }
 
 CAMLprim value caml_ml_channel_size_64(value vchannel)
 {
-  return Val_file_offset(caml_channel_size(Channel(vchannel)));
+  return Val_file_offset(ml_channel_size(vchannel));
 }
 
 CAMLprim value caml_ml_set_binary_mode(value vchannel, value mode)
@@ -731,6 +762,8 @@ CAMLprim value caml_ml_input(value vchannel, value buff, value vstart,
   int n, avail, nread;
 
   Lock(channel);
+ again:
+  check_pending(channel);
   /* We cannot call caml_getblock here because buff may move during
      caml_read_fd */
   start = Long_val(vstart);
@@ -747,6 +780,7 @@ CAMLprim value caml_ml_input(value vchannel, value buff, value vstart,
   } else {
     nread = caml_read_fd(channel->fd, channel->flags, channel->buff,
                          channel->end - channel->buff);
+    if (nread == Io_interrupted) goto again;
     channel->offset += nread;
     channel->max = channel->buff + nread;
     if (n > nread) n = nread;

--- a/runtime/signals.c
+++ b/runtime/signals.c
@@ -62,12 +62,20 @@ CAMLexport int (*caml_sigmask_hook)(int, const sigset_t *, sigset_t *)
   = sigprocmask_wrapper;
 #endif
 
+static int check_for_pending_signals(void)
+{
+  int i;
+  for (i = 0; i < NSIG; i++) {
+    if (caml_pending_signals[i]) return 1;
+  }
+  return 0;
+}
+
 /* Execute all pending signals */
 
 value caml_process_pending_signals_exn(void)
 {
   int i;
-  int really_pending;
 #ifdef POSIX_SIGNALS
   sigset_t set;
 #endif
@@ -78,13 +86,7 @@ value caml_process_pending_signals_exn(void)
 
   /* Check that there is indeed a pending signal before issuing the
      syscall in [caml_sigmask_hook]. */
-  really_pending = 0;
-  for (i = 0; i < NSIG; i++)
-    if (caml_pending_signals[i]) {
-      really_pending = 1;
-      break;
-    }
-  if(!really_pending)
+  if (!check_for_pending_signals())
     return Val_unit;
 
 #ifdef POSIX_SIGNALS
@@ -187,8 +189,10 @@ CAMLexport void caml_leave_blocking_section(void)
      examined by [caml_process_pending_signals_exn], then
      [signals_are_pending] is 0 but the signal needs to be
      handled at this point. */
-  signals_are_pending = 1;
-  //caml_raise_if_exception(caml_process_pending_signals_exn());
+  if (check_for_pending_signals()) {
+    signals_are_pending = 1;
+    caml_set_action_pending();
+  }
 
   errno = saved_errno;
 }

--- a/runtime/signals.c
+++ b/runtime/signals.c
@@ -163,6 +163,11 @@ CAMLexport void caml_enter_blocking_section(void)
   }
 }
 
+CAMLexport void caml_enter_blocking_section_no_pending(void)
+{
+  caml_enter_blocking_section_hook ();
+}
+
 CAMLexport void caml_leave_blocking_section(void)
 {
   int saved_errno;
@@ -183,7 +188,7 @@ CAMLexport void caml_leave_blocking_section(void)
      [signals_are_pending] is 0 but the signal needs to be
      handled at this point. */
   signals_are_pending = 1;
-  caml_raise_if_exception(caml_process_pending_signals_exn());
+  //caml_raise_if_exception(caml_process_pending_signals_exn());
 
   errno = saved_errno;
 }
@@ -320,6 +325,12 @@ Caml_inline value process_pending_actions_with_root_exn(value extra_root)
     CAMLdrop;
   }
   return extra_root;
+}
+
+CAMLno_tsan /* The access to [caml_something_to_do] is not synchronized. */
+int caml_check_pending_actions()
+{
+  return caml_something_to_do;
 }
 
 value caml_process_pending_actions_with_root(value extra_root)

--- a/runtime/signals.c
+++ b/runtime/signals.c
@@ -136,33 +136,18 @@ CAMLno_tsan void caml_record_signal(int signal_number)
 
 /* Management of blocking sections. */
 
-static intnat volatile caml_async_signal_mode = 0;
-
 static void caml_enter_blocking_section_default(void)
 {
-  CAMLassert (caml_async_signal_mode == 0);
-  caml_async_signal_mode = 1;
 }
 
 static void caml_leave_blocking_section_default(void)
 {
-  CAMLassert (caml_async_signal_mode == 1);
-  caml_async_signal_mode = 0;
-}
-
-static int caml_try_leave_blocking_section_default(void)
-{
-  intnat res;
-  Read_and_clear(res, caml_async_signal_mode);
-  return res;
 }
 
 CAMLexport void (*caml_enter_blocking_section_hook)(void) =
    caml_enter_blocking_section_default;
 CAMLexport void (*caml_leave_blocking_section_hook)(void) =
    caml_leave_blocking_section_default;
-CAMLexport int (*caml_try_leave_blocking_section_hook)(void) =
-   caml_try_leave_blocking_section_default;
 
 CAMLno_tsan /* The read of [caml_something_to_do] is not synchronized. */
 CAMLexport void caml_enter_blocking_section(void)

--- a/runtime/signals_byt.c
+++ b/runtime/signals_byt.c
@@ -46,12 +46,7 @@ static void handle_signal(int signal_number)
   signal(signal_number, handle_signal);
 #endif
   if (signal_number < 0 || signal_number >= NSIG) return;
-  if (caml_try_leave_blocking_section_hook()) {
-    caml_raise_if_exception(caml_execute_signal_exn(signal_number, 1));
-    caml_enter_blocking_section_hook();
-  }else{
-    caml_record_signal(signal_number);
-  }
+  caml_record_signal(signal_number);
   errno = saved_errno;
 }
 

--- a/runtime/signals_nat.c
+++ b/runtime/signals_nat.c
@@ -99,19 +99,14 @@ DECLARE_SIGNAL_HANDLER(handle_signal)
   signal(sig, handle_signal);
 #endif
   if (sig < 0 || sig >= NSIG) return;
-  if (caml_try_leave_blocking_section_hook ()) {
-    caml_raise_if_exception(caml_execute_signal_exn(sig, 1));
-    caml_enter_blocking_section_hook();
-  } else {
-    caml_record_signal(sig);
+  caml_record_signal(sig);
   /* Some ports cache [Caml_state->young_limit] in a register.
      Use the signal context to modify that register too, but only if
      we are inside OCaml code (not inside C code). */
 #if defined(CONTEXT_PC) && defined(CONTEXT_YOUNG_LIMIT)
-    if (caml_find_code_fragment_by_pc((char *) CONTEXT_PC) != NULL)
-      CONTEXT_YOUNG_LIMIT = (context_reg) Caml_state->young_limit;
+  if (caml_find_code_fragment_by_pc((char *) CONTEXT_PC) != NULL)
+    CONTEXT_YOUNG_LIMIT = (context_reg) Caml_state->young_limit;
 #endif
-  }
   errno = saved_errno;
 }
 

--- a/runtime/startup_byt.c
+++ b/runtime/startup_byt.c
@@ -444,7 +444,9 @@ CAMLexport void caml_main(char_os **argv)
   /* Load the globals */
   caml_seek_section(fd, &trail, "DATA");
   chan = caml_open_descriptor_in(fd);
+  Lock(chan);
   caml_global_data = caml_input_val(chan);
+  Unlock(chan);
   caml_close_channel(chan); /* this also closes fd */
   caml_stat_free(trail.section);
   /* Ensure that the globals are in the major heap. */

--- a/runtime/unix.c
+++ b/runtime/unix.c
@@ -74,12 +74,13 @@
 int caml_read_fd(int fd, int flags, void * buf, int n)
 {
   int retcode;
-  do {
-    caml_enter_blocking_section();
-    retcode = read(fd, buf, n);
-    caml_leave_blocking_section();
-  } while (retcode == -1 && errno == EINTR);
-  if (retcode == -1) caml_sys_io_error(NO_ARG);
+  caml_enter_blocking_section_no_pending();
+  retcode = read(fd, buf, n);
+  caml_leave_blocking_section();
+  if (retcode == -1) {
+    if (errno == EINTR) return Io_interrupted;
+    else caml_sys_io_error(NO_ARG);
+  }
   return retcode;
 }
 
@@ -92,14 +93,14 @@ int caml_write_fd(int fd, int flags, void * buf, int n)
     retcode = write(fd, buf, n);
   } else {
 #endif
-  caml_enter_blocking_section();
+  caml_enter_blocking_section_no_pending();
   retcode = write(fd, buf, n);
   caml_leave_blocking_section();
 #if defined(NATIVE_CODE) && defined(WITH_SPACETIME)
   }
 #endif
   if (retcode == -1) {
-    if (errno == EINTR) goto again;
+    if (errno == EINTR) return Io_interrupted;
     if ((errno == EAGAIN || errno == EWOULDBLOCK) && n > 1) {
       /* We couldn't do a partial write here, probably because
          n <= PIPE_BUF and POSIX says that writes of less than

--- a/runtime/win32.c
+++ b/runtime/win32.c
@@ -87,7 +87,7 @@ int caml_read_fd(int fd, int flags, void * buf, int n)
 {
   int retcode;
   if ((flags & CHANNEL_FLAG_FROM_SOCKET) == 0) {
-    caml_enter_blocking_section();
+    caml_enter_blocking_section_no_pending();
     retcode = read(fd, buf, n);
     /* Large reads from console can fail with ENOMEM.  Reduce requested size
        and try again. */
@@ -97,7 +97,7 @@ int caml_read_fd(int fd, int flags, void * buf, int n)
     caml_leave_blocking_section();
     if (retcode == -1) caml_sys_io_error(NO_ARG);
   } else {
-    caml_enter_blocking_section();
+    caml_enter_blocking_section_no_pending();
     retcode = recv((SOCKET) _get_osfhandle(fd), buf, n, 0);
     caml_leave_blocking_section();
     if (retcode == -1) caml_win32_sys_error(WSAGetLastError());
@@ -114,7 +114,7 @@ int caml_write_fd(int fd, int flags, void * buf, int n)
     retcode = write(fd, buf, n);
   } else {
 #endif
-    caml_enter_blocking_section();
+    caml_enter_blocking_section_no_pending();
     retcode = write(fd, buf, n);
     caml_leave_blocking_section();
 #if defined(NATIVE_CODE) && defined(WITH_SPACETIME)
@@ -122,7 +122,7 @@ int caml_write_fd(int fd, int flags, void * buf, int n)
 #endif
     if (retcode == -1) caml_sys_io_error(NO_ARG);
   } else {
-    caml_enter_blocking_section();
+    caml_enter_blocking_section_no_pending();
     retcode = send((SOCKET) _get_osfhandle(fd), buf, n, 0);
     caml_leave_blocking_section();
     if (retcode == -1) caml_win32_sys_error(WSAGetLastError());

--- a/testsuite/tests/lib-systhreads/eintr.ml
+++ b/testsuite/tests/lib-systhreads/eintr.ml
@@ -1,0 +1,91 @@
+(* TEST
+
+* hassysthreads
+include systhreads
+** not-windows
+*** bytecode
+*** native
+*)
+
+let signals_requested = Atomic.make 0
+let signal_delay = 0.1
+let _ = Thread.create (fun () ->
+  let signals_sent = ref 0 in
+  ignore (Thread.sigmask Unix.SIG_BLOCK [Sys.sigint]);
+  while true do
+    if Atomic.get signals_requested > !signals_sent then begin
+      Thread.delay signal_delay;
+      Unix.kill (Unix.getpid ()) Sys.sigint;
+      incr signals_sent
+    end else begin
+      Thread.yield ()
+    end
+  done) ()
+let request_signal () = Atomic.incr signals_requested
+
+let () =
+  let (rd, wr) = Unix.pipe () in
+  Sys.catch_break true;
+  request_signal ();
+  begin match Unix.read rd (Bytes.make 1 'a') 0 1 with
+  | _ -> assert false
+  | exception Sys.Break -> print_endline "break: ok" end;
+  Sys.catch_break false;
+  Unix.close rd;
+  Unix.close wr
+
+let () =
+  let (rd, wr) = Unix.pipe () in
+  Sys.set_signal Sys.sigint (Signal_handle (fun _ -> Gc.full_major ()));
+  request_signal ();
+  begin match Unix.read rd (Bytes.make 1 'a') 0 1 with
+  | _ -> assert false
+  | exception Unix.Unix_error(Unix.EINTR, "read", _) ->
+     print_endline "eintr: ok" end;
+  Sys.set_signal Sys.sigint Signal_default;
+  Unix.close rd;
+  Unix.close wr
+
+
+(* Doing I/O on stdout would be more realistic, but seeking has the
+   same locking & scheduling effects, without actually producing any
+   output *)
+let poke_stdout () =
+  match out_channel_length stdout with
+  | _ -> ()
+  | exception Sys_error _ -> ()
+
+let () =
+  let r = Atomic.make true in
+  Sys.set_signal Sys.sigint (Signal_handle (fun _ ->
+    poke_stdout (); Atomic.set r false));
+  request_signal ();
+  while Atomic.get r do
+    poke_stdout ()
+  done;
+  Sys.set_signal Sys.sigint Signal_default;
+  print_endline "chan: ok"
+
+let () =
+  let mklist () = List.init 1000 (fun i -> (i, i)) in
+  let before = Sys.opaque_identity (ref (mklist ())) in
+  let during = Atomic.make (Sys.opaque_identity (mklist ())) in
+  let siglist = ref [] in
+  Sys.set_signal Sys.sigint (Signal_handle (fun _ ->
+    Gc.full_major (); poke_stdout (); Gc.compact ();
+    siglist := mklist ();
+    raise Sys.Break));
+  request_signal ();
+  begin match
+    while true do
+      poke_stdout ();
+      Atomic.set during (mklist ())
+    done
+  with
+  | () -> assert false
+  | exception Sys.Break -> () end;
+  let expected = Sys.opaque_identity (mklist ()) in
+  assert (!before = expected);
+  assert (Atomic.get during = expected);
+  assert (!siglist = expected);
+  print_endline "gc: ok"

--- a/testsuite/tests/lib-systhreads/eintr.reference
+++ b/testsuite/tests/lib-systhreads/eintr.reference
@@ -1,0 +1,4 @@
+break: ok
+eintr: ok
+chan: ok
+gc: ok

